src/main/java/duke/command/Command.java

```diff
@@ -0,0 +1,21 @@
+package duke.command;
+
+import duke.exception.DukeException;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+import java.io.IOException;
+
+/**
+ * An abstract class to enforce a structure
+ * for command classes.
+ */
+public abstract class Command {
+
+    public abstract void execute(TaskList taskList, Ui ui, Storage storage)

```

Comment:
Perhaps you can consider throwing DukeExceptions only and have messages in these exceptions be displayed ui.showError?